Fix Xaml code generator to account for access modifier

Youssef1313 opened the pull request:
GitHub Issue (If applicable): #5411

<!-- Link to relevant GitHub issue if applicable. All PRs should be associated with an issue (GitHub issue or internal) -->

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Code is always generated with `public` modifier causing a compile-time error.

## What is the new behavior?

Code is now generated with the correct modifier so the code compiles successfully.

## PR Checklist

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below.
     Please note that breaking changes are likely to be rejected -->

## Other information

Fixes #5411

Internal Issue (If applicable):

jeromelaban commented on src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs:
```diff
@@ -371,8 +376,13 @@ private string InnerGenerateFile()
 					AnalyzerSuppressionsGenerator.Generate(writer, _analyzerSuppressions);
 
 					var controlBaseType = GetType(topLevelControl.Type);
+					var symbol = _generatorContext.Compilation.GetTypeByMetadataName($"{_className.ns}.{_className.className}");
+					if (symbol is null)
+                    {
+						throw new Exception($"Symbol '{_className.ns}.{_className.className}' was not found.");
+                    }
```
`GetType` (like the one immediately above) is doing this validation and raising an exception. Is there something that you saw not working with it ?

Youssef1313 replied:
@jeromelaban I haven't noticed the method. Updated the code. Thanks!

Youssef1313 commented:
Just knew that not adding an access modifier at all is okay for the compiler and is different than an explicit "internal":

```csharp
// This doesn't cause any errors
public partial class C1 {}
partial class C1 {}

// This causes an error:
// error CS0262: Partial declarations of 'C2' have conflicting accessibility modifiers
public partial class C2 {}
internal partial class C2 {}
```

I simplified the fix in 4566a1d

jeromelaban commented:
Change is looking good! Can you squash your history so it's only one commit ?

Also, can you check if it's supported in UWP to create a XAML with an x:Class which does not exist in code behind? This change will try to get a type that does not exist (because it'll be generated). If UWP does not allow for it, it'll be fine. Otherwise we'll have to make the `GetType` a `FindType` and ignore the accessibility if the symbol can't be found.

Youssef1313 commented:
@jeromelaban Having an `x:Class="App3.MainPage"` in a new UWP app doesn't produce any errors if `MainPage` class doesn't exist. Looking at `obj\x86\Debug`, it gets generated in two files as follows:

**MainPage.g.cs**:

```csharp
#pragma checksum "C:\Users\PC\source\repos\App3\App3\MainPage.xaml" "{8829d00f-11b8-4213-878b-770e8597ac16}" "5621C471234A6B1A7370EE0BE1DEF59EDE123537C614199A43E5C781E666FA91"
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

namespace App3
{
    partial class MainPage : 
        global::Windows.UI.Xaml.Controls.Page, 
        global::Windows.UI.Xaml.Markup.IComponentConnector,
        global::Windows.UI.Xaml.Markup.IComponentConnector2
    {
        /// <summary>
        /// Connect()
        /// </summary>
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Windows.UI.Xaml.Build.Tasks"," 10.0.19041.685")]
        [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
        public void Connect(int connectionId, object target)
        {
            this._contentLoaded = true;
        }

        /// <summary>
        /// GetBindingConnector(int connectionId, object target)
        /// </summary>
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Windows.UI.Xaml.Build.Tasks"," 10.0.19041.685")]
        [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
        public global::Windows.UI.Xaml.Markup.IComponentConnector GetBindingConnector(int connectionId, object target)
        {
            global::Windows.UI.Xaml.Markup.IComponentConnector returnValue = null;
            return returnValue;
        }
    }
}
```

**MainPage.g.i.cs**:

```csharp
#pragma checksum "C:\Users\PC\source\repos\App3\App3\MainPage.xaml" "{8829d00f-11b8-4213-878b-770e8597ac16}" "5621C471234A6B1A7370EE0BE1DEF59EDE123537C614199A43E5C781E666FA91"
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

namespace App3
{
    partial class MainPage : global::Windows.UI.Xaml.Controls.Page
    {


        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Windows.UI.Xaml.Build.Tasks"," 10.0.19041.685")]
        private bool _contentLoaded;

        /// <summary>
        /// InitializeComponent()
        /// </summary>
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Windows.UI.Xaml.Build.Tasks"," 10.0.19041.685")]
        [global::System.Diagnostics.DebuggerNonUserCodeAttribute()]
        public void InitializeComponent()
        {
            if (_contentLoaded)
                return;

            _contentLoaded = true;

            global::System.Uri resourceLocator = new global::System.Uri("ms-appx:///MainPage.xaml");
            global::Windows.UI.Xaml.Application.LoadComponent(this, resourceLocator, global::Windows.UI.Xaml.Controls.Primitives.ComponentResourceLocation.Application);
        }

        partial void UnloadObject(global::Windows.UI.Xaml.DependencyObject unloadableObject);

    }
}



```

Does this require any changes? If yes, can you please elaborate more? Thanks!

jeromelaban commented:
So as UWP does not require an actual code-behind class, this means that:

```csharp
var symbol = GetType($"{_className.ns}.{_className.className}");
```

Will incorrectly fail because the type will not be available in the `Compilation` instance, and this this not what UWP does.

It should be:

```csharp
var symbol = FindType($"{_className.ns}.{_className.className}");
```
You can add a validation for that in 
 https://github.com/unoplatform/uno/tree/master/src/SourceGenerators/XamlGenerationTests.

Youssef1313 commented:
@jeromelaban Ooh I have removed that line from GitHub UI in 4566a1d, but the change was lost when I squashed locally without pulling the change from GitHub first. Will fix!
